### ISSUE TITLE
An issue when opening multiple databases.

### ISSUE THREAD

**Issue by serhii-k** [tekartik/sqflite#276]
Hello Alex,

First of all thanks so much for your wonderful plugin!

I've noticed an issue on Android. In my app I have several databases with the same structure (e.g. for several users who can use the app on the same device). And it's possible to switch a user -- the app will simply work with another database.

Important: I don't close databases (even on the exit, I don't know how to call the dispose when the app is closed).

Let's simplify things. We have only 2 users. For each user we have a database. Each database contains only 1 table. For the first user we have file user1.db, and for the second user2.db.

user1.db contains 1 table with 6 records
user2.db contains 1 table without records

We have a List < Database > with open connections. 

When we first start the app, it works smoothly.
1) At the beginning the first user is initialized, the app opens the connection to the first db, and reads 6 records from its table.
2) Then we switch to the second user, the app opens the connection to the second db, and reads 0 records from its table.
3) We can switch to the first user. The app uses already open connection, and reads 6 records from it.
4) We switch to the second user. The app uses another open connection, and reads 0 records.
And so on...

Here is where the issue happens.
We press the "Back" button to close the app. And then we open the app again (warm restart).
1) Let's say we have stopped on the first user (we store the last user). The app opens the connection to the user1.db, and reads 6 records from its table.
2) We switch to the second user. The app opens the connection to user2.db (yes, for sure). BUT! Instead of 0 records, it reads 6 same records!
3) And so on... Both connections are open, and the app still reads 6 records again, and again.
It looks like the app at this stage opens only the first used database, and uses it for all subsequent connections.

If we stop on the 2nd user and do the warm restart, the app reads 0 records from both connections.

After the cold restart (we swipe away the app from the list of recent apps), the app runs smoothly again, and opens two specified database files.

One remark: I've tried to close the previous user database when we switch to a different user. And this issue didn't appear in this case.

Please can you think about this case. Maybe I just have to close connections, but it really looks like some bug (as the app works correctly after a cold restart). Thank you!

sqflite: I've tried 1.1.5 and 1.1.6
Android Emulator (API 27)

**Comment by alextekartik** [comment]
Thanks for the report. Indeed that's definitely an issue. It will take some time to reproduce on my side though. sqflite has some special trick to handle this "warm restart" on the native side so maybe it does messup if there are multiple databases.

Before I look into it (not likely before 1 or 2 weeks), can you try the following by adding `Sqflite.devSetDebugModeOn(true);` at the beginning of your  `main()`. It is deprecated on purpose to avoid keeping it in your code. It should log your sql query executed and I think it also display the database id for each query so maybe it could help figuring out what is happening (you can filter using grep and mabye post your logs here when you perform your 'back' scenario).

**Comment by serhii-k** [comment]
Yes sure, Alex.

Here is the log with some comments.

I added some prints to this log:
1) The method getDb() is based on your example with await _lock.synchronized()
(https://github.com/tekartik/sqflite/blob/master/sqflite/doc/opening_db.md)
I added prints to check _db is null, and around the openDatabase().
2) And also I logged the number of returned records.

It looks when the issue appears, **openDatabase** method doesn't open the second database.

```
Initial run. The current User is 1.
15:03:47.747 28237-28278/com.package I/flutter:    getDb() for user 1: _db == null ? true (just to make sure we don't have an open connection)
15:03:48.226 28237-28278/com.package I/flutter:    getDb() for user 1: before openDatabase(/data/user/0/com.package/databases/user1.db)
15:03:48.261 28237-28237/com.package D/Sqflite: starting threadThread[Sqflite,5,main] priority 10
15:03:48.261 28237-28237/com.package D/Sqflite: [Thread[main,5,main]] opened 1 /data/user/0/com.package/databases/user1.db
15:03:48.270 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] opened 1 /data/user/0/com.package/databases/user1.db
15:03:49.126 28237-28300/com.package D/Sqflite: [1,Sqflite(822)] BEGIN EXCLUSIVE
15:03:49.141 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] PRAGMA user_version
15:03:49.191 28237-28300/com.package D/Sqflite: [1,Sqflite(822)] COMMIT
15:03:49.198 28237-28278/com.package I/flutter:    getDb() for user 1: after openDatabase(/data/user/0/com.package/databases/user1.db)
15:03:49.211 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] SELECT * FROM scores_avg7
15:03:49.225 28237-28278/com.package I/flutter:    read 6 records
We switch to User 2.
15:04:14.928 28237-28278/com.package I/flutter:    getDb() for user 2: _db == null ? true (just to make sure we don't have an open connection)
15:04:14.955 28237-28278/com.package I/flutter:    getDb() for user 2: before openDatabase(/data/user/0/com.package/databases/user2.db)
15:04:14.963 28237-28237/com.package D/Sqflite: [Thread[main,5,main]] opened 2 /data/user/0/com.package/databases/user2.db
15:04:14.969 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] opened 2 /data/user/0/com.package/databases/user2.db
15:04:14.974 28237-28300/com.package D/Sqflite: [2,Sqflite(822)] BEGIN EXCLUSIVE
15:04:14.988 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] PRAGMA user_version
15:04:14.997 28237-28300/com.package D/Sqflite: [2,Sqflite(822)] COMMIT
15:04:15.000 28237-28278/com.package I/flutter:    getDb() for user 2: after openDatabase(/data/user/0/com.package/databases/user2.db)
15:04:15.002 28237-28300/com.package D/Sqflite: [Thread[Sqflite,5,main]] SELECT * FROM scores_avg7
15:04:15.011 28237-28278/com.package I/flutter:    read 0 records
Back and Warm Restart.
The current User is 2.
15:04:27.777 28237-28338/com.package I/flutter:    getDb() for user 2: _db == null ? true (just to make sure we don't have an open connection)
15:04:29.087 28237-28338/com.package I/flutter:    getDb() for user 2: before openDatabase(/data/user/0/com.package/databases/user2.db)
15:04:29.112 28237-28237/com.package D/Sqflite: starting threadThread[Sqflite,5,main] priority 10
15:04:29.113 28237-28237/com.package D/Sqflite: [Thread[main,5,main]] opened 1 /data/user/0/com.package/databases/user2.db
15:04:29.120 28237-28343/com.package D/Sqflite: [Thread[Sqflite,5,main]] opened 1 /data/user/0/com.package/databases/user2.db
15:04:29.142 28237-28343/com.package D/Sqflite: [1,Sqflite(827)] BEGIN EXCLUSIVE
15:04:29.195 28237-28343/com.package D/Sqflite: [Thread[Sqflite,5,main]] PRAGMA user_version
15:04:29.225 28237-28343/com.package D/Sqflite: [1,Sqflite(827)] COMMIT
15:04:29.322 28237-28338/com.package I/flutter:    getDb() for user 2: after openDatabase(/data/user/0/com.package/databases/user2.db)
15:04:29.338 28237-28343/com.package D/Sqflite: [Thread[Sqflite,5,main]] SELECT * FROM scores_avg7
15:04:29.349 28237-28338/com.package I/flutter:    read 0 records
We switch to User 1.
15:04:34.039 28237-28338/com.package I/flutter:    getDb() for user 1: _db == null ? true (just to make sure we don't have an open connection)
15:04:34.100 28237-28338/com.package I/flutter:    getDb() for user 1: before openDatabase(/data/user/0/com.package/databases/user1.db)
15:04:34.114 28237-28343/com.package D/Sqflite: [1,Sqflite(827)] BEGIN EXCLUSIVE
15:04:34.117 28237-28343/com.package D/Sqflite: [Thread[Sqflite,5,main]] PRAGMA user_version
15:04:34.121 28237-28343/com.package D/Sqflite: [1,Sqflite(827)] COMMIT
15:04:34.135 28237-28338/com.package I/flutter:    getDb() for user 1: after openDatabase(/data/user/0/com.package/databases/user1.db)
15:04:34.138 28237-28343/com.package D/Sqflite: [Thread[Sqflite,5,main]] SELECT * FROM scores_avg7
15:04:34.153 28237-28338/com.package I/flutter:    read 0 records
```

**Comment by alextekartik** [comment]
Thanks @serhii-k I was able to pinpoint some issue in this recover database scenario on the java side. I was able to test the scenario you provided in the example app (manual test | multiple databases) by trying with 3 different databases and testing the various warm restart, hot restart...I think this is fixed in the version `1.1.7-dev.1` I just published. I would appreciate if you could try it before I release an official new version. Thanks!

**Comment by serhii-k** [comment]
Oh, wow! It works!!! :-)
Alex, thank you so much for your help!

**Comment by alextekartik** [comment]
Great. Thanks for the taking the time to fill the issue and testing the fix! I publish 1.1.7 (same as 1.1.7-dev.1).

**Comment by serhii-k** [comment]
I'm really glad to be of help :)
